### DISCUSSION TITLE
Resistors and ESP32 ADC

### DISCUSSION THREAD

**Discussion by Lonestar78** [nliaudat/esp32_8ch_motor_shield#19] [General]
Dear @nliaudat,

quick one....i use v54, rev 1.34....there i find 1,07 Ohm resistors between BEMF and Ground (R3).
Measuring BEMF on the open header gives me ~10mv if a valve opens/closes without load, ~30mV on endstop (open and klack klack klack...).
Given that the ESP32 adc reads 0 for 0-0,1v how do you get ADC readings?
i always get 0....damn, can it be, that the 1,06 Ohm Resistor is to small? Why did you change it from 12 Ohm?
Attenuation is set to 2.5db and 12bit resolution.

Any hint is welcome. 
May it be, that i have to re-solder bigger (= higher ohm) to R3?
also could use 10kOhm there and another 10kOhm between header for BEMF and VCC.
What do you think?

Thanks!
Christian

**Comment by nstadigs** [comment]
Greetings!

1.07 ohm resistors don't work for me either. See https://github.com/nliaudat/esp32_8ch_motor_shield/issues/9.
I resoldered 12 ohm resistors to my boards and that works well for me. I don't know why @nliaudat settled for 1.07 after that discussion, but since 1.07 seems to work for others, I have a working solution myself and the fact that I don't understand half of these things I just dropped it. :man_shrugging:

**Comment by Lonestar78** [comment]
It might depend on the esp32.

The S2 version reads from 0 volt, not only from 0,1...

Holen Sie sich Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by nliaudat** [comment]
Hi,
L9110s chips have a buildin resistance between gnd and bemf which is undocumented. 12 ohms Did not work for some actuators, that's why I change it to the smallest available 1.07ohm.

You can add a resistor in open headers between gnd and bemf to fit your needs. I do not remember if it setup serial or parallel.

**Reply by nliaudat** [reply]
The soldered 1.07 ohm resistors are in parallel and divid themselfs (1.07 || 1.07 = 0.54 per channel ). 

Il you add resistors in empty headers, they are in series and add themselfs.
You could try adding small resistors in empty headers, without removing the builtin one

**Reply by Lonestar78** [reply]
OK, i removed on CH1 + CH2 the two 1,07Ohm resistors and replaced with 2x 10Ohm on the openheader between the two BEMF pins and GND, this seems to work.

**Reply by Tecnologic** [reply]
Same issue here. With 10Ohms my valves have no torque at all. The 1R Resistor ist fine but the ADC of the ESP32 is crap it cant sense any thing below 80mV so i will add a simple opamp to get a suitable voltage for the ADC. External ADCs like ADS1115 will work to but are more expensive. 
![schematic for Opamp](https://user-images.githubusercontent.com/1442404/191107917-10470adb-64b1-4b04-8ac8-c58732554e7a.jpg)
![current to voltage vor adc](https://user-images.githubusercontent.com/1442404/191107943-3c04569d-292c-4052-a9d3-11cf57d2c1fa.jpg)

Here u see with that simple opamp u get a voltage of 800mV for 0mA and 80mV for 100mA Gain could be changed to fit everyones need.

Another thing why the hell are there no 100nF decoupling Caps on that board?

**Reply by Lonestar78** [reply]
using an ADS1115 is no bad idea at all....especially since i have a couple ;-)

**Reply by Lonestar78** [reply]
From a board design perspective, the opamp certainly makes sense. This part of the board should be made more robust. Not my turf though.
Or a sum amplifier, which allways adds 100mV....

**Reply by Tecnologic** [reply]
The inverting one with offset is the least complicated. But I will start with ads1115 because the rework with these breakout boards is simple and u gain extra channels.

**Reply by nliaudat** [reply]
@Tecnologic sorry not to implement 100nf capacitors at ADC reading inputs, but the reason was the adc sensor update interval is 500ms, so no need to have a bypass capacitor for such very slow resolution. There are 100nf caps before motor

 _10Ohms my valves have no torque_ : What kind a power supply do you use ? Is it on all channels ?

**Reply by Tecnologic** [reply]
@nliaudat I mean 100nF at each supply pin pair of every device on the Board at least. 

The ADC of the esp is so bad that it will not gain any better readings with it a cap there I'm with you.

I use vd-mod drives and just a single channel at once. Because I couldn't get the current endstops to trigger with the stock 1R.  I don't see any changes in the ADC output for a fully blocked drive with 1R. For me the blocking current is roughly 100mA so 50mV on the shunt. With the ADC only starting at 80mV the measurement is useless. I will switch to piggy back ads1115 boards and stick with the 1R resistor.

If I find the time to get the eda files to kicad I do a redesign.

**Reply by nliaudat** [reply]
I have 3 testing boards with different esp32 and all detect 100mv. As sample, my moving valve consume approx 150mv and the endstops are set between 250-300mv. I mean you sould have a bad power supply cause the ADC depends strictly on the Vref input. It could also explain why you have no torque with 10ohm resistors. Regards

**Reply by Tecnologic** [reply]
My drives are rated 120mA with the 1R || 1R thats 60mV and thats What i see. The PSU is the LCSC 230V to 3V3 brick. From the looks of the 3V3 on my Scope its doing its job. With 10R || 10R the motors stop because it looks like the L91xx is not getting enough voltage to switch any more. Current wise i have 500mV there and the ADC redings of the ESP32 reflect that but the drive has no torque anymore. With the 1R variant i did no current measurements up until now. 

To get a OK measurenment i think changing the ADC or adding OPAMPs is the only way to go. Then the cover with current endstops template should work as expected.

**Reply by nliaudat** [reply]
L9110 chip can work with 2.5V at 200uA. Have you a laboratory power supply for testing ?

**Reply by Tecnologic** [reply]
Sure, will test in the evening

**Reply by nliaudat** [reply]
Any news ? Have you tested with external power supply ?

**Reply by Tecnologic** [reply]
No had no time. Hopefully today evening

**Reply by flexmatics** [reply]
> OK, i removed on CH1 + CH2 the two 1,07Ohm resistors and replaced with 2x 10Ohm on the openheader between the two BEMF pins and GND, this seems to work.

Confirm, that this fix has also worked for me.

**Reply by nethawk2** [reply]
I use v54, rev 1.34 and replaced the eight 1,07 Ohm resistors by 10 Ohm resistors too. Now BEMF works, if i set BEMF trigger to 0.20.
My first SMD soldering ...

**Reply by nliaudat** [reply]
> I use v54, rev 1.34 and replaced the eight 1,07 Ohm resistors by 10 Ohm resistors too. Now BEMF works, if i set BEMF trigger to 0.20. My first SMD soldering ...

No need to remove the 1,07o resistors, you can add some in the free headers labeled BEMF to GND
![146636319-be3378fe-499a-4a70-8dc2-b36edf379c23](https://user-images.githubusercontent.com/6782613/193602318-5f3a0570-ce97-4066-835a-4b3fb14e61c5.jpg)

**Comment by Tecnologic** [comment]
I now hat time to post my findings. My HmIP-VdMot motors are rated on the Label with 120mA max. and 3V3 supply voltage. During running the draw 20mA while opening and 30mA while closing. When they run to block while opening they draw roughly 40mA before the click click noise (gear teeth slipping) is coming. When closing they draw 120mA before the drive pops up from the valve. 

So i switched to dual ads1115 as adc for all channels and used analog_threshold at 40mA for opening and 80mA for closing and they work like a charm.  

![PXL_20220922_134251076 MP](https://user-images.githubusercontent.com/1442404/192114670-4bf889f2-e42f-4600-a839-0e8429592861.jpg)

**Reply by herrfrei** [reply]
@Tecnologic: What is the update interval you reach with the two boards on i2c? Could you please provide your yaml config?

**Reply by Tecnologic** [reply]
Hi, 
This is my yaml https://github.com/Tecnologic/floor-heating-controller

**Reply by herrfrei** [reply]
Hi, it looks that you are not using the ads1115 anymore and developed a new board. May I ask you, why?

**Reply by Tecnologic** [reply]
I have problems with the reliability of the values endpositons. Some valves do not completely close occasionally. Some times they completely block end I need to adjust the threshold and so on. It is running right now and working ok. But I need to monitor and maintain it every month. 

From my point of view it's the simple current peak endstop mechanism which is flawed. I do industrial motor control for a living and can't sustain such a solution. But with esphome it's difficult to get this better with pure yaml.

**Comment by nliaudat** [comment]
ads1115 smt chip costs 5.42$ at JLCPCB ==> +11$ price increase for the board.

**Reply by Tecnologic** [reply]
The adafruit breakout boards are 3€ a piece. But for a PCB redisign I would use the opamp circuit I posted above with an gs8631 opamp (lcsc stock) and a cd74hc4067 analog multiplexer.

The cd74hc4067 is supported by esphome and with filter settings for the sensors one can do the offset and inversion of the opamp signal to get a valid current reading. 

By the way I tried the easy eda links to join the design but I have no rights to make changes. Need some guidance there. Other wise I would use kicad.

**Reply by nliaudat** [reply]
Many thanks, I've set you as developper in easyeda (default is observer)

**Reply by Tecnologic** [reply]
Thanks, don't expect anything next week, hole lot of work that week.

**Reply by nstadigs** [reply]
Hi! Has there been any progress on this? Or should I get me some ads1115's?

**Reply by Tecnologic** [reply]
For now I run two units with ads1115 without flaws. It's just a bit ugly

**Comment by Tecnologic** [comment]
For now I run two units with ads1115 without flaws. It's just a bit ugly

**Comment by nliaudat** [comment]
For information, the 1R resistor for BEMF works great with ESP32-C3. 
I've done some tests with https://github.com/nliaudat/floor-heating-proportional-valve and I can read from 0.00v (not 0.08V) with steps of 0.01v correctly. 
I mean that the problem of reading BEMF is related with some different quality of ESP32-WROOM-32D.

**Reply by nstadigs** [reply]
I think I've read somewhere that the ESP32-Sx and ESP32-Cx models have better ADC measuring capabilities than the standard ESP32. I guess that would explain the different attenuation voltage ranges in the docs:

- ESP32: https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/peripherals/adc.html
- ESP32-C3: https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32c3/api-reference/peripherals/adc.html

**Reply by nstadigs** [reply]
If this is true, maybe we should be looking at moving to an S-model instead of changing the circuit for reading the BEMF?

Both the S2 and S3 models have the same frequency as the ESP32-WROOM-32E ones but the S3 ones are a too expensive I think. It seems to cost twice as much as the S2. So maybe we should try the S2? The S2 does not have bluetooth, but we don't use that here anyway.

Here are the ADC docs for the S2: https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32s2/api-reference/peripherals/adc.html

What are your thoughts, @nliaudat , @Tecnologic ?

**Reply by nstadigs** [reply]
https://www.aliexpress.com/item/1005003578116249.html?spm=a2g0o.productlist.0.0.769e2332lRDqik&algo_pvid=de3fd4b8-7ffe-4e23-b77d-49e928d1bac1&algo_exp_id=de3fd4b8-7ffe-4e23-b77d-49e928d1bac1-41&pdp_ext_f=%7B%22sku_id%22%3A%2212000026353623793%22%7D&pdp_npi=2%40dis%21SEK%21217.83%21141.55%21%21%21%21%21%402100bdde16683345752768109e7812%2112000026353623793%21sea&curPageLogUid=qXJpNtXSOZKz

It's only single-core, but I don't think we need a sleep mode, since it's not battery powered.

**Reply by Tecnologic** [reply]
Thats a good solution with minimal effort. I'm currently looking in to running the actuators current controlled because they are quite unreliable when running a pid_climate. But this would be possible with the ads1115 or the s2 esp2.

**Reply by nstadigs** [reply]
I'm not sure if the S2 has the same adc vs wifi quirks as the ESP32-WROOM-32E. If we're lucky we could even use all the adc-channels of the S2. Can't find any information about this tho...

EDIT:

From the docs:
> The ESP32-S2 integrates 2 SAR ([Successive Approximation Register](https://en.wikipedia.org/wiki/Successive_approximation_ADC)) ADCs, supporting a total of 20 measurement channels (analog enabled pins).

ADC1 and ADC2 have 10 channels each so maybe that does not even matter?

**Reply by nstadigs** [reply]
> I'm currently looking in to running the actuators current controlled because they are quite unreliable when running a pid_climate. But this would be possible with the ads1115 or the s2 esp2.

Very interesting :)

**Reply by nliaudat** [reply]
S2, S3 and C3 devkits do not have the same pinout as ESP32-wroom-32D.

The only full compatible is [ESP-C3-32S](https://docs.ai-thinker.com/_media/esp32/docs/esp-c3-32s_%E8%8B%B1%E6%96%87.pdf) which is no anymore for sell

**Reply by nstadigs** [reply]
Yes, I understand that. I was thinking that instead of adding components to the circuit for reading BEMF as Tecnologic suggested we could modify the board to fit the pinout of the S2 devkit. This would at least require less components.

**Reply by nliaudat** [reply]
changing the devkit is not so easy as pinout :
- The S2 & C3 compile only with esp-idf framework and that means the web server is not available anymore.
- S2 have ADC1 10 channels: GPIO1 - GPIO10 free (ADC2 is always used by wifi) = ok but the board is expensive (~10$)
- C3 have ADC1 5 channels: GPIO0 - GPIO4 free = ok but expensive (ESP32-C3-DevKitM-1 = ~ 12$)

There are differences between espressif documentation and devkit. As example, the wroom-32D has 8 ADC1 channels, but only 6 are available (ACD1_CH0 and ACD1_CH3 to ACD1_CH7)  while ADC1_CH1 and ADC1_CH2 are not available (even the pins are not exposed in the ESP32 Development Board).

Changing of devkit must be carefully studied and the price raise from 3.5$ to 10-12$ for the devkit
Unfortunately, I have just changed jobs and I no longer have any free time.

**Reply by nstadigs** [reply]
I see. Thanks for the explanation. Maybe we should just go the opamp route proposed by Tecnologic then?

Good luck with your new job!

**Reply by nliaudat** [reply]
> Thats a good solution with minimal effort. I'm currently looking in to running the actuators current controlled because they are quite unreliable when running a pid_climate. But this would be possible with the ads1115 or the s2 esp2.


@Tecnologic 
I have tested pid climate but the very high update frequency was bad for valve life. 
I've used the PID_mode: kp: 30 ki: 0.005 kd: -24000 for my low temperature underfloor heating.

Have you find good parameters ?

**Reply by Tecnologic** [reply]
OK. Set KD to 0. Thats a floor heating that is just a big pool sink for energy. There is nothing the Controller needs to compensate quickly. Then determine the rate of ur Temperature Update. My is 60s. Then assume the time constant of ur room. Typical Something like 8h. Then U choose how much the valve should react for a 1K Error in Temperature, i would start with 0.03 (1=100%) for 1K. So Kp would be 0.03. With the time constant of 8h Set as TN that gives a ki = Kp / (TN / UpdateRate) = 0.03/(8h*3600s/60s) = 0.0000625

**Comment by nliaudat** [comment]
For information, esphome support now the 8 channels [adc128s102](https://esphome.io/components/sensor/adc128s102.html).
It cost ~4.2$ (much more than the ESP32 devkit)

**Comment by nliaudat** [comment]
@Tecnologic Do you mind adding a LM358 opamp chip could fix the ADC reading ?

**Reply by Tecnologic** [reply]
Yes. U need one amp per Channel so a Quad amp or an an analog multiplexer Like i suggested

**Comment by Karl-opec** [comment]
A Quad Op-Amp LM324 cost less then 10cent at JLCPCB some Resitors and capacitors to add, would be a cheap solution to fix the ADC-Readings.

**Comment by nliaudat** [comment]


**Reply by Karl-opec** [reply]
First, for my understanding, what is the reason to have the 1.07 Ohm Resistor two times per channel block(CH1,Ch2)? These resistors are in parallel, which halves the resulting value (0.5 ohm). So a 1ohm resistor would be sufficient.  (Power 625 µW)

The Voltage over the resistor is a measure for the motor current. The blocking Motor current according to this Graph [Vd-mot Motor current](https://github.com/Lenti84/VdMot_Controller/blob/master/system/actuators.md) is 25mA. Over a 1Ohm resistor this lead to 25mV to Ground as CH(1,2)-BEMF. so the Gain of the OP-Amps should be greater than 5 to have a voltage over 100mV for the ESP32 ADC.

Am I right or do i miss something?
The capacitor in the output of the op-amp makes BEMF to ESP32 ADC AC-coupled. I think we need it DC-coupled. 

These are my very general comments. My time for having done such design are long time away, Perhaps   @Tecnologic can make comments more valid.

**Reply by nliaudat** [reply]
Many thanks for you support !

I am not an expert on electronics, but the L9100s motor driver is a bit uncommon ... If you remove the 1R resistor, it do not works well. 
Some people need to cut them off and replace with 12Ω resistors in series to get it works. (but not everybody, it seems to be random, adding 12Ω resistors works without cutting 1R)
My own board works great since 3 years with 1R resistors and ADC is very good. I mean there is factory ADC calibration problem with some cheap boards. 

My homematic motors run between 0.01A to 0.024A and endstop is at 0.08A-0.1A. It differs from motor ...

I have no experiences with opAmp, that's why I ask. 
You are right for the capacitors. They are useless.

![Capture d’écran 2022-11-28 230853](https://user-images.githubusercontent.com/6782613/204391467-462a4fcf-1520-4418-8a49-86dbfe633ac9.jpg)


[Easyeda link](https://easyeda.com/editor#id=|240c2bd91cde438f93348d56e1ae4e72|420e6f6085d643fc9c5df7bfbe9595bf|f15c181c211e4aebaf86420464abe718|b3e9b48180db4901b45d0292a792846e|141370bd57bc41da8cc555929a203303|60a1b0936f664eb8b9e5f7402068a21b|4da0648c1a03403c84598c3102a931dd|5fcc1c941d5c41d2a409d14e1d0494dd|c35090f023e0471a9a0e92ded8f16f99|b7d74c27dc73434a9bbc4ada70b343df|ddeaacb3e13c40f3886b31e9330730b1|7744c3e75c54448eb9fed788f130dd96)

**Reply by Karl-opec** [reply]
The Shunt Resitor from L9110s GND pin to electric GND of the circuit is definitly needed. You need it to generate a Voltage which is a measure for the motor current.  The Motor Current is what you want to measure . The unknown value of an internal resistor is of no interest.

**Comment by nliaudat** [comment]
@Tecnologic @Lonestar78 @Karl-opec : Can you validate the [schematic for LM324 opamp](https://easyeda.com/editor#id=|240c2bd91cde438f93348d56e1ae4e72|420e6f6085d643fc9c5df7bfbe9595bf|f15c181c211e4aebaf86420464abe718|b3e9b48180db4901b45d0292a792846e|141370bd57bc41da8cc555929a203303|60a1b0936f664eb8b9e5f7402068a21b|4da0648c1a03403c84598c3102a931dd|5fcc1c941d5c41d2a409d14e1d0494dd|c35090f023e0471a9a0e92ded8f16f99|b7d74c27dc73434a9bbc4ada70b343df|ddeaacb3e13c40f3886b31e9330730b1|7744c3e75c54448eb9fed788f130dd96) ?

![image](https://user-images.githubusercontent.com/6782613/205343343-52f915ef-324d-4bf6-8092-bd3f71cac2ad.png)

![image](https://user-images.githubusercontent.com/6782613/205346998-353be8f0-427d-4404-bc8f-5c5f98b72e32.png)

calculation done at [http://mustcalculate.com/electronics/noninvertingopamp.php?vin=0.01&r1=10k&r2=50k](http://mustcalculate.com/electronics/noninvertingopamp.php?vin=0.01&r1=10k&r2=50k)

**Reply by Tecnologic** [reply]
Hi was very busy week. I would spend a extra 1k and 1n low pass with the 1n cap placed right at the adc pin of the esp. That gives the adc a good low impedence signal for sampling.

Mean while im trying a different approach: https://github.com/Tecnologic/floor-heating-controller-s2
My goal is still to make "servo" drives out of the valves. But ich think i can not stick to esphome with that approach.

**Reply by nliaudat** [reply]
many thanks for your contribution. 
I see your L9110s implementation and do not undertand everything. I am an IT admin, not an electronics.
This looks very promising !

![image](https://user-images.githubusercontent.com/6782613/205375159-e1ee266c-7160-4860-9a79-9422e4e17316.png)

**Reply by Karl-opec** [reply]
I have JLCPCB let build me  the PCB v55 rev 1.45. It works very good so far. I have replaced the 1.07 R with 2 ohms. so in parallel it is 1 ohm.
With the gain of the op-amp (6) this gives blocking voltages at 25mA / 1 Ohm * 6 = 0.150V. setting these values as bemf_trigger it works great. I propose you add also the low pass-filter which @Tecnologic mentioned and it would be a additional improvement. The resistor issue would be solved.

You could save in addition 4 Resitors, if you leave off one of the shunts per channel. 1 Ohm is sufficiant. There are not two needed.

**Reply by nliaudat** [reply]
Many thanks. 
Feel free to commit your improvements in Easyeda [V56 Karl-opec improvements](https://easyeda.com/editor#id=|240c2bd91cde438f93348d56e1ae4e72|420e6f6085d643fc9c5df7bfbe9595bf|f15c181c211e4aebaf86420464abe718|b3e9b48180db4901b45d0292a792846e|60a1b0936f664eb8b9e5f7402068a21b|7744c3e75c54448eb9fed788f130dd96|78ed6c97623e4c1f929c753245e2f96b)
I have no time actually to do - new job. sorry

**Reply by xdubx** [reply]
did the v56 needs a new label? On the pcb stands still "PCB v55 rev 1.45"

**Reply by nliaudat** [reply]
v56 is the same as v55. No time to do the changes

**Reply by nstadigs** [reply]
> I have JLCPCB let build me the PCB v55 rev 1.45. It works very good so far. I have replaced the 1.07 R with 2 ohms. so in parallel it is 1 ohm. With the gain of the op-amp (6) this gives blocking voltages at 25mA / 1 Ohm * 6 = 0.150V. setting these values as bemf_trigger it works great. I propose you add also the low pass-filter which @Tecnologic mentioned and it would be a additional improvement. The resistor issue would be solved.
> 
> You could save in addition 4 Resitors, if you leave off one of the shunts per channel. 1 Ohm is sufficiant. There are not two needed.

@Karl-opec You don't happen to have a copy of your BOM and pick-and-place files do you? :)

**Reply by Karl-opec** [reply]
@nstadigs : you find them under the above link.

In the BOM  I have replaced  the 1,07 Resistor against 2.0 ohm resistors . NR. @ JLCPCB  C17942 Manufacturer Nr. 1206W4F200KT5E

**Comment by xdubx** [comment]
I have one from azdelivery and the problem with the 0.08v BEMF to. 
Try the next days my other ESP32 i hope one of them works fine. 

The c und S2 version working without problems?  Can someone recommend a shop where you are sure not to get a buggy esp.

**Reply by xdubx** [reply]
OK i double checked the electrical voltages. One of my valves vom homematic have weird value.

**Comment by nliaudat** [comment]
New version v57 rev 1.47 :

    Replace the 1.07R with 2 ohms

**Comment by nliaudat** [comment]
The opamp do not work in latest version. 
I do not understand why ...

**Reply by WMP** [reply]
I am supposed to receive 10 pieces of V57 rev 1.48 in a few days, can I help you somehow and test something?

**Reply by nliaudat** [reply]
I have received 5 last week, but the BEMF values reported by the opamp are the same as without. It should have a gain of 6x. It works, but without amplification.

**Reply by nliaudat** [reply]
I tested to inject 0.2V directly in the BEMF pin and I get 1.2V as output. The opamp works, but if I connect the esp32, the AC/DC converter and the Vmot motors, I get only the previous BEMF values near 0.3V.

**Reply by Karl-opec** [reply]
@nliaudat getting close to 0.3 V ist absolutliy fine. U(bemf) = R * I(mot) *  V(opamp) -> 0,3V = 1ohm* 0,05A *6 .

R(shunt) CH1 || R(shunt) CH2. 2ohm parallel 2Ohm -> 1ohm resulting.

50 mA Motor Current ist Blocking motor.  I have setup in the software as Motor Endstop Voltage 0,180 V. since then it work . No more Pop up's from the valve etc.  Look at this [chart](https://github.com/Lenti84/VdMot_Controller/blob/master/system/actuators.md) to see the Motorcurrent over valveopenening.

If you want higher values of  U(bemf) you can increase V. But this not recommend becaus noise is also increased.

following find my definition of sensor_adc.yaml:
```
sensor:
### in use the motor use 15-20 mV. At endstop, it raise to 50 mV
  - platform: adc
    pin: GPIO36
    name: "BEMF 1" #CH1 + CH2
    update_interval: 100ms
    attenuation : 0db
    id : BEMF_1_2_sensor_ADC
    accuracy_decimals: 3
    filters:
    - sliding_window_moving_average:
        window_size: 9
        send_every: 9

```
This averages the voltage measurment of 900ms und gives a value every 900ms.

this works for me since 4 Month now

**Reply by nliaudat** [reply]
@Karl-opec many thanks for your sensor_adc.yaml sharing. I've tested it for fast an hour with a small modification to get 0.000V instead of 0.075

```
    filters:
    - offset: -0.075
```

My BEMF trigger is set to : 

`  ch1_bemf_trigger: "0.105"`
(0.180-0.075) instead of 0.180

It seems to works.

BUT : I do not understant why my previous BEMF trigger value (0.200 to 0.350 depending the valve) without opamp is the same or less with opamp.

**Reply by nliaudat** [reply]
@Karl-opec : do you want  I add you as developper ? You have more skills than me in electronic and could improve the project.

**Comment by Karl-opec** [comment]
Thank you for you offer, yes it would be nice. 

your previous BEMF trigger can be reached. look at the Motor Current chart. While Blocking the motor can gain higher currents and this is leading to high BEMF Trigger Values. But remember you want to measure the Motor Current. and more than 30-35 mA are not needed to trigger.

because of different blocking currents for openning and closing i have split the trigger levels for open and close.

the use of the filter is not realy of any use. as the trigger level is the switch pont.

**Reply by nliaudat** [reply]
What is your easyeda member name ? 
You are invited in github as collaborator

**Reply by Karl-opec** [reply]
my Username is Karl23

**Comment by nliaudat** [comment]
Everybody who's want to contribute must say. 
I have closed the public easyeda sharing cause nobody commited an improvement since 3 years and it was a rant

The main goal is to make an open source heating controller for everybody. The only contributors are in that discussion and you are welcome to improve or get more involved.

Regards and many thanks for your improvements

**Comment by nliaudat** [comment]
Here is the easyeda link to view the project : https://oshwlab.com/project/join/b2642362910a768d05dcb075a611a4d2
I'll add everybody to the project as developper again if you want to contribute.
